restore config locks

dguibert opened the pull request:
fixes https://github.com/flatironinstitute/nixpack/issues/40

With https://github.com/spack/spack/commit/53ae969aa008c8fb358f33b9b961beb5512999ff, config:locks is no longer used to create Lock. This patch restores this and allow to create database locks to a custom directory.

Database locks have been introduced by https://github.com/spack/spack/commit/53ae969aa008c8fb358f33b9b961beb5512999ff
We could either use config:locks on them or create the lock in a custom directory. To avoid much trouble, the later has been chosen.

@alalazo @haampie

alalazo reviewed (changes requested):
Can you explain what is the issue you are facing in Spack? 

This PR doesn't make much sense to me, since it reads (again) from a global variable values that should be injected in the constructor.  If there is a value that is wrong for the store or DB etc. it should be fixed where the object is created, not in the class constructor by reading a global variable.

alalazo commented:
> With https://github.com/spack/spack/commit/53ae969aa008c8fb358f33b9b961beb5512999ff, config:locks is no longer used to create Lock. 

Why do you say this? See:

https://github.com/spack/spack/blob/af8c392de2a5c7c1b1ca1448964dbd4bab87c43a/lib/spack/spack/database.py#L345-L351

which creates a `LockConfiguration` from a `Configuration` object and:

https://github.com/spack/spack/blob/af8c392de2a5c7c1b1ca1448964dbd4bab87c43a/lib/spack/spack/store.py#L209-L233

which uses it to construct the store. The PR you mention had the only purpose of decoupling the internal design. Before the PR we couldn't construct a lock without having a global configuration object available.

alalazo commented on lib/spack/spack/store.py:
```diff
@@ -266,7 +267,8 @@ def _construct_upstream_dbs_from_install_roots(
         upstream_dbs = list(accumulated_upstream_dbs)
-        next_db = spack.database.Database(
+        db_root = spack.config.get("config:database_root", install_root)
+        next_db = spack.database.Database(db_root,
```
> The CI fails with: "Database" gets multiple values for keyword argument "upstream_dbs" [misc]
> But it's not introduced by this patch.

Are you sure?

dguibert replied:
sorry, missed that

alalazo commented:
> I think that DataBase and FileCache objects are not instantiated with the global config:locks setting, right?

No, they should be instantiated like that - and they generally are. It might be though that some code path got an accidental bug during the refactor, so I'm trying to understand that. 

The modification interposes a factory layer that translates the global `config:locks` settings in:

https://github.com/spack/spack/blob/af8c392de2a5c7c1b1ca1448964dbd4bab87c43a/lib/spack/spack/store.py#L236-L238

where `spack.config.config` is the global configuration contaning `config:locks`.

alalazo commented:
> I'm facing this issue https://github.com/flatironinstitute/nixpack/issues/40, paths are read-only, locks are not needed.

@dguibert It would be great if you could extract a Spack only reproducer for that. That would help me a lot triaging and eventually fixing the bug.

dguibert committed:
make misc_cache obey config:locks

dguibert committed:
database: use config:database_root, root as fallback

In nixpack, root (defined as install_tree:root) is defined to a non
existing path. Causing error when creating database directory
(Line 617 of lib/spack/spack/database.py).

This patch allows to define a specific root for database: config:database_root

alalazo commented:
Any luck trying to get a reproducer in Spack?
